File: scala/raindrops/src/main/scala/raindrops.scala
class Raindrops {

  import Raindrops.RaindropsUtils

  def convert(number: Int): String = {
    val melody = Raindrops.Drops
      .keys
      .filter(_.isFactorOf(number))
      .map(Raindrops.Drops(_))

    if (melody.isEmpty) {
      number.toString
    } else {
      melody.mkString
    }
  }

}

object Raindrops {

  def apply(): Raindrops = new Raindrops()

  val Drops: Map[Int, String] = Map(
    (3 -> "Pling"),
    (5 -> "Plang"),
    (7 -> "Plong")
  )

  implicit class RaindropsUtils(val factor: Int) {
    def isFactorOf(number: Int): Boolean = number % factor == 0
  }

}

File: java/raindrops/src/main/java/Raindrops.java
import java.util.stream.Collectors;
import java.util.Map;
import java.util.HashMap;
import java.util.LinkedHashMap;

public class Raindrops {

  private static Map<Integer, String> DROPS = new LinkedHashMap<>();
  static {
    DROPS.put(3, "Pling");
    DROPS.put(5, "Plang");
    DROPS.put(7, "Plong");
  }

  public static String convert(int number) {
    String sound = DROPS.entrySet().stream()
      .filter(entry -> number % entry.getKey() == 0)
      .map(entry -> entry.getValue())
      .collect(Collectors.joining());

    if (sound.isEmpty()) {
      return String.valueOf(number);
    } else {
      return sound;
    }
  }
}